### PR TITLE
[collector] Add ability to enable GOMEMLIMIT

### PR DESCRIPTION
**Pull request by TylerHelmuth**
This PR starts the potential deprecation processor of the memory ballast extension. 

A new flag, `useGOMEMLIMIT` is added that, when enabled, removes the memory ballast extension from the config and adds a `GOMEMLIMIT` env var.

Added a new CI scenario and a new example.

Related to https://github.com/open-telemetry/opentelemetry-helm-charts/issues/891 and https://github.com/open-telemetry/opentelemetry-collector/issues/8343

### CODE REVIEW

File: charts/opentelemetry-collector/UPGRADING.md [hunk 1]
```diff
@@ -4,6 +4,16 @@ These upgrade guidelines only contain instructions for version upgrades which re
 If the version you want to upgrade to is not listed here, then there is nothing to do for you.
 Just upgrade and enjoy.
 
+## something to soemthing
```
[the remark this answers is not in the document]
Comment (TylerHelmuth) [reply charts/opentelemetry-collector/UPGRADING.md]: woops ya

File: charts/opentelemetry-collector/templates/NOTES.txt [hunk 2]
```diff
@@ -48,3 +48,7 @@
 {{- if and (eq .Values.dnsPolicy "None") (not .Values.dnsConfig) }}
 {{- fail "[ERROR] dnsConfig should be provided when dnsPolicy is None" }}
 {{ end }}
+
+{{- if and (.Values.useGOMEMLIMIT) (not ((((.Values.resources).limits).memory))) }}
+[WARNING] "useGOMEMLIMIT" is enabled but memory limits have not been supplied. It is best to use "useGOMEMLIMIT" when memory limits have been set.
```
Comment (mx-psi) [inline charts/opentelemetry-collector/templates/NOTES.txt:53]: nit: be more explicit that `useGOMEMLIMIT` does nothing if the resource limit is not set.

**Review by mx-psi — APPROVED** [review]

**Comment by TylerHelmuth** [comment]
@dmitryax please review

File: charts/opentelemetry-collector/templates/NOTES.txt [hunk 3]
```diff
@@ -31,6 +31,14 @@
 [WARNING] Setting internalTrafficPolicy to 'Cluster' on Daemonset is not recommended. Consider using 'Local' instead.
 {{ end }}
 
+{{- if and (eq .Values.dnsPolicy "None") (not .Values.dnsConfig) }}
+{{- fail "[ERROR] dnsConfig should be provided when dnsPolicy is None" }}
+{{ end }}
```
Comment (dmitryax) [inline charts/opentelemetry-collector/templates/NOTES.txt:36]: This seems unrelated

**Review by dmitryax — APPROVED** [review]

File: charts/opentelemetry-collector/templates/_config.tpl [hunk 4]
```diff
@@ -20,10 +20,19 @@ Merge user supplied config into memory limiter config.
 {{- if not $processorsConfig.memory_limiter }}
 {{-   $_ := set $processorsConfig "memory_limiter" (include "opentelemetry-collector.memoryLimiter" . | fromYaml) }}
 {{- end }}
-{{- $memoryBallastConfig := get .Values.config.extensions "memory_ballast" }}
-{{- if or (not $memoryBallastConfig) (not $memoryBallastConfig.size_in_percentage) }}
-{{-   $_ := set $memoryBallastConfig "size_in_percentage" 40 }}
+
+{{- if and .Values.useGOMEMLIMIT }}
```
Comment (jinja2) [inline charts/opentelemetry-collector/templates/_config.tpl:24]:
```suggestion
{{- if .Values.useGOMEMLIMIT }}
```

File: charts/opentelemetry-collector/templates/_pod.tpl [hunk 5]
```diff
@@ -47,6 +47,10 @@ containers:
           fieldRef:
             fieldPath: spec.nodeName
       {{- end }}
+      {{- if and (.Values.useGOMEMLIMIT) ((((.Values.resources).limits).memory))  }}
+      - name: GOMEMLIMIT
+        value: {{ div (mul (include "opentelemetry-collector.convertMemToMib" .Values.resources.limits.memory) 80) 100 }}Mi
```
Comment (jinja2) [inline charts/opentelemetry-collector/templates/_pod.tpl:52]:
```suggestion
        value: {{ div (mul (include "opentelemetry-collector.convertMemToMib" .Values.resources.limits.memory) 80) 100 }}MiB
```

`The supported suffixes include B, KiB, MiB, GiB, and TiB.` from the [docs](https://pkg.go.dev/runtime).

File: charts/opentelemetry-collector/templates/NOTES.txt [hunk 6]
```diff
@@ -31,6 +31,10 @@
 [WARNING] Setting internalTrafficPolicy to 'Cluster' on Daemonset is not recommended. Consider using 'Local' instead.
 {{ end }}
 
+{{- if and (.Values.useGOMEMLIMIT) (not ((((.Values.resources).limits).memory))) }}
+[WARNING] "useGOMEMLIMIT" is enabled but memory limits have not been supplied, which means no GOMEMLIMIT env var was configured but the Memory Ballast Extension was removed. It highly recommended to only use "useGOMEMLIMIT" when memory limits have been set.
```
Comment (jinja2) [inline charts/opentelemetry-collector/templates/NOTES.txt:35]:
```suggestion
[WARNING] "useGOMEMLIMIT" is enabled but memory limits have not been supplied, which means no GOMEMLIMIT env var was configured but the Memory Ballast Extension was removed. It is highly recommended to only use "useGOMEMLIMIT" when memory limits have been set.
```
